autofit/graphical/messages/transform.py

```diff
@@ -0,0 +1,339 @@
+
+
+class FunctionTransform(AbstractDensityTransform):
+    def __init__(self, func, inv_func, grad, hess=None, args=(), func_grad_hess=None):
+        self.func = func
+        self.inv_func = inv_func
+        self.grad = grad
+        self.hess = hess
+        self.args = args
+        self.func_grad_hess = func_grad_hess
+
+    def transform(self, x):
+        return self.func(x, *self.args)
+
+    def inv_transform(self, x):
+        return self.inv_func(x, *self.args)
+
+    def jacobian(self, x):
+        return DiagonalMatrix(self.grad(x, *self.args))
+
+    def log_det(self, x: np.ndarray) -> np.ndarray:
+        gs = self.grad(x, *self.args)
+        return np.log(gs)
+
+    def log_det_grad(self, x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        if self.func_grad_hess:
+            x0, gs, hs = self.func_grad_hess(x, *self.args)
+        else:
+            x0 = self.func(x, *self.args)
+            gs = self.grad(x, *self.args)
+            hs = self.hess(x, *self.args)
+        return np.log(gs), hs/gs
+
+    def transform_det_jac(
+        self, x
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, LinearOperator]:
+        if self.func_grad_hess:
+            x0, gs, hs = self.func_grad_hess(x, *self.args)
+        else:
+            x0 = self.func(x, *self.args)
+            gs = self.grad(x, *self.args)
+            hs = self.hess(x, *self.args)
+        return x0, np.log(gs), hs/gs, DiagonalMatrix(gs)
+
+
+def exp3(x):
+    expx = np.exp(x)
+    return (expx, ) * 3
+
+
+exp_transform = FunctionTransform(np.exp, np.log, np.exp, func_grad_hess=exp3)
+
+
+def log3(x):
+    ix = np.reciprocal(x)
+    return np.log(x), ix, -np.square(ix)
+
+
+log_transform = FunctionTransform(
+    np.log, np.exp, np.reciprocal, func_grad_hess=log3)
+
+
+def sigmoid(x, scale=1, shift=0):
+    return scale / (1 + np.exp(-x)) + shift
+
+
+def logit(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    return np.log(x) - np.log1p(-x)
+
+
+def sigmoid_grad(x, scale=1, shift=0):
+    expx = np.exp(-x)
+    return scale * expx / np.square(1 + expx)
+
+
+def logit_grad(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    return (np.reciprocal(x) + np.reciprocal(1 - x)) / scale
+
+
+def logit_hess(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    return np.reciprocal(1-x) - np.reciprocal(x)
+
+
+def logit_grad_hess(x, scale=1, shift=0):
+    x = (x - shift) / scale
+    ix = np.reciprocal(x)
+    ix1 = np.reciprocal(1 - x)
+    ix2 = np.square(ix)
+    ix12 = np.square(ix1)
+    return (np.log(x) - np.log1p(-x), (ix + ix1)/scale, (ix12 - ix2)/scale**2)
+
+
+logistic_transform = FunctionTransform(
+    logit, sigmoid, logit_grad, func_grad_hess=logit_grad_hess)
+
+
+def shifted_logistic(shift=0, scale=1):
+    return FunctionTransform(logit, sigmoid, logit_grad, func_grad_hess=logit_grad_hess, args=(scale, shift))
+
+
+def ndtri_grad(x):
+    return np.reciprocal(_norm_pdf(ndtri(x)))
+
+def ndtri_hess(x):
+    return
```

Comment:
lol that's a very consequential function.

Reply:
oops